plugins/DRTubesSDAction.cpp

```diff
@@ -0,0 +1,267 @@
+  // ---
+  // dd4hep::BitFieldCoder
+  // decoder("stave:10,tower:6,air:1,col:16,row:16,clad:1,core:1,cherenkov:1");
+  // auto VolID = volumeID(aStep); auto CherenkovID = decoder.get(VolID,"cherenkov");
+  // ---
+  // However using regexSensitiveDetector does not populate the cache that allows using the
+  // volumeID() method. One can still access volIDs in this sensitive detector action with something
+  // like this
+  // ---
+  // G4VPhysicalVolume* PhysVol = aStep->GetPreStepPoint()->GetTouchableHandle()->GetVolume();
+  // Geant4Mapping& Mapping = Geant4Mapping::instance();
+  // PlacedVolume PlacedVol = Mapping.placement(PhysVol);
+  // const PlacedVolumeExtension::VolIDs& TestIDs = PlacedVol.volIDs();
+  // auto it = TestIDs.find("name");
+  // std::cout<<it->first<<" "<<it->second<<std::endl;
+  // ---
+  // but this brute force method makes the simulaiton slower by more than two order of magnitudes
+  // (see https://github.com/AIDASoft/DD4hep/issues/1319).
+  // Therefore we use copynumbers instead of volIDs.
+
+#ifdef DRTubesSDDebug
+  // Print out some info step-by-step in sensitive volumes
+  //
+  DRTubesSglHpr::PrintStepInfo(aStep);
+#endif
+
+  auto Edep = aStep->GetTotalEnergyDeposit();
+  auto cpNo = aStep->GetPreStepPoint()->GetTouchable()->GetCopyNumber();
+  // The second bit of the CopyNumber corresponds to the "core" entry:
+  // 1 if the step is in the fiber core (S or C) and 0 if it is
+  // in the fiber cladding (C only) 
+  unsigned int CoreID = (cpNo & 0b10) >> 1; // take CpNo 2nd bit
+  bool IsCherClad = (CoreID == 0);
+  // The first bit of the CopyNumber corresponds to the "cherenkov" entry
+  // 1 for C fibers and 0 for S fibers
+  unsigned int CherenkovID = cpNo & 0b1;
+  bool IsScin = (CherenkovID == 0);
+  // Skip this step if edep is 0 and it is a scintillating fiber
+  if (IsScin && Edep == 0.) return true;
+  // If it is a track inside the cherenkov CLADDING skip this step,
+  // if it is an optical photon kill it first
+  if (IsCherClad) {
+    if (aStep->GetTrack()->GetParticleDefinition() == G4OpticalPhoton::Definition()) {
+      aStep->GetTrack()->SetTrackStatus(fStopAndKill);
+    }
+    return true;
+  }
+
+  // Now we are inside fibers' core volume (either Scintillating or Cherenkov)
+
+  // We recreate the TubeID from the tube copynumber:
+  // fist16 bits for the columnID and second 16 bits for the rowID
+  auto TubeID = aStep->GetPreStepPoint()->GetTouchable()->GetCopyNumber(2);
+  unsigned int ColumnID = TubeID >> 16;
+  unsigned int RawID = TubeID & 0xFFFF;
+  auto TowerID = static_cast<unsigned int>(aStep->GetPreStepPoint()->GetTouchable()->GetCopyNumber(3));
+  auto StaveID = static_cast<unsigned int>(aStep->GetPreStepPoint()->GetTouchable()->GetCopyNumber(4));
+
+  VolumeID VolID = 0; // recreate the 64-bit VolumeID
+  BitFieldCoder bc("stave:10,tower:6,air:1,col:16,row:16,clad:1,core:1,cherenkov:1");
```

Comment:
If this has to match with the readout `<id>` this should have a lot of comments in the XML